File: cube_master.py
import argparse
import json
import os
from pathlib import Path
from typing import Any

import requests


BASE_URL = "https://api.cubemaster.net"


def parse_response(resp: requests.Response) -> Any:
	content_type = resp.headers.get("Content-Type", "")
	if "application/json" in content_type:
		return resp.json()
	return resp.text


def call_api(
	method: str,
	path: str,
	token: str,
	params: dict[str, Any] | None = None,
	payload: dict[str, Any] | None = None,
) -> tuple[int, Any]:
	url = f"{BASE_URL}{path}"
	headers = {"TokenID": token}

	resp = requests.request(
		method=method.upper(),
		url=url,
		headers=headers,
		params=params,
		json=payload,
		timeout=30,
	)
	return resp.status_code, parse_response(resp)


def print_result(status: int, body: Any) -> None:
	print(f"HTTP {status}")
	if isinstance(body, (dict, list)):
		print(json.dumps(body, ensure_ascii=False, indent=2))
	else:
		print(body)


def command_check_token(token: str) -> None:
	status, body = call_api("GET", "/Loads", token, params={"limit": 1})
	print_result(status, body)


def command_list_loads(token: str, limit: int) -> None:
	status, body = call_api("GET", "/Loads", token, params={"limit": limit})
	print_result(status, body)


def command_get_load(token: str, load_id: str) -> None:
	status, body = call_api("GET", f"/Loads/{load_id}", token)
	print_result(status, body)


def command_create_load(token: str, payload_file: str) -> None:
	payload_path = Path(payload_file)
	if not payload_path.exists():
		raise SystemExit(f"Payload file not found: {payload_file}")

	with payload_path.open("r", encoding="utf-8") as f:
		payload = json.load(f)

	status, body = call_api("POST", "/Loads", token, payload=payload)
	print_result(status, body)


def command_call_path(token: str, method: str, path: str, limit: int) -> None:
	params: dict[str, Any] | None = None
	if path == "/Loads" and method.upper() == "GET":
		params = {"limit": limit}

	status, body = call_api(method, path, token, params=params)
	print_result(status, body)


def build_parser() -> argparse.ArgumentParser:
	parser = argparse.ArgumentParser(description="CubeMaster API integration helper")
	parser.add_argument(
		"--token",
		default="",
		help="API TokenID. If omitted, reads CUBEMASTER_TOKEN.",
	)

	subparsers = parser.add_subparsers(dest="command")

	check_parser = subparsers.add_parser("check-token", help="Step 1: validate token with GET /Loads")
	check_parser.add_argument("--token", default="", help="Override token for this command")

	list_parser = subparsers.add_parser("list-loads", help="Step 2: list load results")
	list_parser.add_argument("--token", default="", help="Override token for this command")
	list_parser.add_argument("--limit", type=int, default=1, help="Number of records to fetch")

	get_parser = subparsers.add_parser("get-load", help="Step 4: query one load by id")
	get_parser.add_argument("--token", default="", help="Override token for this command")
	get_parser.add_argument("--load-id", required=True, help="Load id to query")

	create_parser = subparsers.add_parser("create-load", help="Step 3: create calculation via POST /Loads")
	create_parser.add_argument("--token", default="", help="Override token for this command")
	create_parser.add_argument("--payload-file", required=True, help="Path to JSON payload file")

	path_parser = subparsers.add_parser("call", help="Call any API path directly")
	path_parser.add_argument("--token", default="", help="Override token for this command")
	path_parser.add_argument("--method", default="GET", help="HTTP method")
	path_parser.add_argument("--path", required=True, help="API path, for example /Loads")
	path_parser.add_argument("--limit", type=int, default=1, help="limit for GET /Loads")

	return parser


def main() -> None:
	parser = build_parser()
	args = parser.parse_args()
	token = args.token or os.getenv("CUBEMASTER_TOKEN", "")

	if not token:
		raise SystemExit("Missing token: pass --token or set CUBEMASTER_TOKEN")

	if args.command in (None, "check-token"):
		command_check_token(token)
	elif args.command == "list-loads":
		command_list_loads(token, args.limit)
	elif args.command == "get-load":
		command_get_load(token, args.load_id)
	elif args.command == "create-load":
		command_create_load(token, args.payload_file)
	elif args.command == "call":
		command_call_path(token, args.method, args.path, args.limit)
	else:
		parser.print_help()


if __name__ == "__main__":
	main()

File: cube_master.js
import { readFileSync } from 'fs';
import { resolve } from 'path';

const BASE_URL = 'https://api.cubemaster.net';

async function callApi(method, path, token, { params, payload } = {}) {
  let url = `${BASE_URL}${path}`;
  if (params) {
    url += `?${new URLSearchParams(params).toString()}`;
  }
  const init = {
    method: method.toUpperCase(),
    headers: { 'TokenID': token },
    signal: AbortSignal.timeout(30000),
  };
  if (payload !== undefined) {
    init.headers['Content-Type'] = 'application/json';
    init.body = JSON.stringify(payload);
  }
  const resp = await fetch(url, init);
  const contentType = resp.headers.get('content-type') || '';
  const body = contentType.includes('application/json')
    ? await resp.json()
    : await resp.text();
  return { status: resp.status, body };
}

function printResult(status, body) {
  console.log(`HTTP ${status}`);
  if (typeof body === 'object' && body !== null) {
    console.log(JSON.stringify(body, null, 2));
  } else {
    console.log(body);
  }
}

function parseArgs(argv) {
  const args = argv.slice(2);
  const result = { token: '', command: null, opts: { limit: 1, method: 'GET' } };
  for (let i = 0; i < args.length; i++) {
    if      (args[i] === '--token')        result.token = args[++i];
    else if (args[i] === '--limit')        result.opts.limit = parseInt(args[++i], 10);
    else if (args[i] === '--load-id')      result.opts.loadId = args[++i];
    else if (args[i] === '--payload-file') result.opts.payloadFile = args[++i];
    else if (args[i] === '--method')       result.opts.method = args[++i];
    else if (args[i] === '--path')         result.opts.path = args[++i];
    else if (!args[i].startsWith('--') && !result.command) result.command = args[i];
  }
  if (!result.token) result.token = process.env.CUBEMASTER_TOKEN || '';
  return result;
}

function printHelp() {
  console.log(`CubeMaster API integration helper

Usage:
  node cube_master.js <command> --token <TOKEN> [options]

Commands:
  check-token                        Step 1: validate token with GET /Loads
  list-loads    [--limit N]          Step 2: list load results (default: 1)
  create-load   --payload-file FILE  Step 3: create calculation via POST /Loads
  get-load      --load-id ID         Step 4: query one load by id
  call          --method M --path P  Call any API path directly

Options:
  --token TOKEN   API TokenID (or set CUBEMASTER_TOKEN env var)
`);
}

async function main() {
  const { token, command, opts } = parseArgs(process.argv);

  if (!token) {
    console.error('Missing token: pass --token or set CUBEMASTER_TOKEN');
    process.exit(1);
  }

  if (!command || command === 'check-token') {
    const { status, body } = await callApi('GET', '/Loads', token, { params: { limit: 1 } });
    printResult(status, body);

  } else if (command === 'list-loads') {
    const { status, body } = await callApi('GET', '/Loads', token, { params: { limit: opts.limit } });
    printResult(status, body);

  } else if (command === 'get-load') {
    if (!opts.loadId) { console.error('--load-id is required'); process.exit(1); }
    const { status, body } = await callApi('GET', `/Loads/${opts.loadId}`, token);
    printResult(status, body);

  } else if (command === 'create-load') {
    if (!opts.payloadFile) { console.error('--payload-file is required'); process.exit(1); }
    const filePath = resolve(opts.payloadFile);
    let payload;
    try {
      payload = JSON.parse(readFileSync(filePath, 'utf-8'));
    } catch {
      console.error(`Cannot read payload file: ${filePath}`);
      process.exit(1);
    }
    const { status, body } = await callApi('POST', '/Loads', token, { payload });
    printResult(status, body);

  } else if (command === 'call') {
    if (!opts.path) { console.error('--path is required'); process.exit(1); }
    let params;
    if (opts.path === '/Loads' && opts.method.toUpperCase() === 'GET') {
      params = { limit: opts.limit };
    }
    const { status, body } = await callApi(opts.method, opts.path, token, { params });
    printResult(status, body);

  } else {
    printHelp();
  }
}

main().catch(err => {
  console.error(err.message);
  process.exit(1);
});
